### DISCUSSION TITLE
handle multiple form of function content

### DISCUSSION THREAD

**Discussion by setop** [ast-grep/ast-grep#1279] [Q&A]
Thanks for this amazing tool ! 

Here is what I experiment:

# intro

In Elixir, dependencies are declared in a `mix.exs` file, in a function, usually named `deps` that returns a list of tuples.

```elixir
defmodule Pleroma.Mixfile do
  use Mix.Project
  #...

  defp deps do
    [
      {:phoenix, "~> 1.7.3"},
      {:phoenix_ecto, "~> 4.4"},
      {:ecto_sql, "~> 3.10"},
      {:postgrex, ">= 0.0.0"}
    ]
  end

  #...
end
```

if I want to declare a new dependency I can do it with [this](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImVsaXhpciIsInF1ZXJ5IjoiY29uc29sZS5sb2coJE1BVENIKSIsInJld3JpdGUiOiJsb2dnWFhYWFhYWFhYZXIubG9nKCRNQVRDSCkiLCJjb25maWciOiJsYW5ndWFnZTogRWxpeGlyXG5ydWxlOlxuICBwYXR0ZXJuIDogfFxuICAgIGRlZnAgZGVwcyBkbyBbICQkJEEgXSBlbmRcbmZpeDogfFxuICBkZWZwIGRlcHMgZG9cbiAgICBbXG4gICAgICAkJCRBLFxuICAgICAgezpodHRwX3NpZ25hdHVyZXMsIFwifj4gMC4xLjJcIn1cbiAgICBdXG4gIGVuZCIsInNvdXJjZSI6ImRlZm1vZHVsZSBQbGVyb21hLk1peGZpbGUgZG9cbiAgdXNlIE1peC5Qcm9qZWN0XG4gICMuLi5cblxuICBkZWZwIGRlcHMgZG9cbiAgICBbXG4gICAgICB7OnBob2VuaXgsIFwifj4gMS43LjNcIn0sXG4gICAgICB7OnBob2VuaXhfZWN0bywgXCJ+PiA0LjRcIn0sXG4gICAgICB7OmVjdG9fc3FsLCBcIn4+IDMuMTBcIn0sXG4gICAgICB7OnBvc3RncmV4LCBcIj49IDAuMC4wXCJ9XG4gICAgXVxuICBlbmRcblxuICAjLi4uXG5lbmRcbiJ9)

```yaml
language: Elixir
rule:
  pattern: defp deps do [ $$$A ] end
fix: |
  defp deps do
    [
      $$$A,
      {:http_signatures, "~> 0.1.2"}
    ]
  end
```

And it can be repeated over and over to add more dependencies.

The most complicated is to put the comma at the end of the last line. It is very fragile with `diff/patch`, it is trivial with `asg-grep`.

Very nice. Love it.

# issue

Unfortunatly, since it is a function, the list can be build in many ways, especially, on some projects, the function is concatenating the results of other functions:

```elixir
  defp deps do
    [
      {:phoenix, "~> 1.7.3"},
      {:phoenix_ecto, "~> 4.4"},
      {:ecto_sql, "~> 3.10"},
      {:postgrex, ">= 0.0.0"}
    ] ++ logger_deps()
  end
```

This helps to organize the dependencies or even to make them depending on some properties of the environment.

But then the pattern I use does not work anymore. I have to use [this](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImVsaXhpciIsInF1ZXJ5IjoiY29uc29sZS5sb2coJE1BVENIKSIsInJld3JpdGUiOiJsb2dnWFhYWFhYWFhYZXIubG9nKCRNQVRDSCkiLCJjb25maWciOiJsYW5ndWFnZTogRWxpeGlyXG5ydWxlOlxuICBwYXR0ZXJuIDogfFxuICAgIGRlZnAgZGVwcyBkbyBbICQkJEEgXSArKyAkJCRCIGVuZFxuZml4OiB8XG4gIGRlZnAgZGVwcyBkb1xuICAgIFtcbiAgICAgICQkJEEsXG4gICAgICB7Omh0dHBfc2lnbmF0dXJlcywgXCJ+PiAwLjEuMlwifVxuICAgIF0gKysgJCQkQlxuICBlbmQiLCJzb3VyY2UiOiJkZWZtb2R1bGUgUGxlcm9tYS5NaXhmaWxlIGRvXG4gIHVzZSBNaXguUHJvamVjdFxuICAjLi4uXG5cbiAgZGVmcCBkZXBzIGRvXG4gICAgW1xuICAgICAgezpwaG9lbml4LCBcIn4+IDEuNy4zXCJ9LFxuICAgICAgezpwaG9lbml4X2VjdG8sIFwifj4gNC40XCJ9LFxuICAgICAgezplY3RvX3NxbCwgXCJ+PiAzLjEwXCJ9LFxuICAgICAgezpwb3N0Z3JleCwgXCI+PSAwLjAuMFwifVxuICAgIF0gKysgbG9nZ2VyX2RlcHMoKVxuICBlbmRcblxuICAjLi4uXG5lbmRcbiJ9).

Which is not compatible with the simple form above and not compatible with a more complicated one like `++ logger_deps() ++ auth_deps()` ...

# question 

I'd like something more generic and more robust, like "in the root module, there is a function 'deps', find the first list, insert my tuple, no matter what the rest of the function looks like.

I tried with `inside` and `has` but was quickly lost.

Any hint ?

**Comment by HerringtonDarkholme** [comment]
https://ast-grep.github.io/advanced/find-n-patch.html `rewriter` should be ideal to solve your problem.

https://ast-grep.github.io/reference/yaml/rewriter.html You can also take a loot at rewriter's doc here.
